services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaServiceITest.kt

```diff
@@ -210,6 +208,95 @@ class HankeKayttajaServiceITest : DatabaseTest() {
             )
     }
 
+    @Test
+    fun `saveNewTokensFromHanke does nothing if hanke has no contacts`() {
+        hankeKayttajaService.saveNewTokensFromHanke(HankeFactory.create())
+
+        assertThat(kayttajaTunnisteRepository.findAll()).isEmpty()
+        assertThat(hankeKayttajaRepository.findAll()).isEmpty()
+    }
+
+    @Test
+    fun `saveNewTokensFromHanke creates tokens for unique ones`() {
+        val hanke =
+            hankeFactory.save(
+                HankeFactory.create()
+                    .withYhteystiedot(
+                        // each has a duplicate
+                        omistajat = listOf(1, 1),
+                        rakennuttajat = listOf(2, 2),
+                        toteuttajat = listOf(3, 3),
+                        muut = listOf(4, 4)
+                    )
+            )
+        assertThat(hanke.extractYhteystiedot()).hasSize(8)
+
+        hankeKayttajaService.saveNewTokensFromHanke(hanke)
+
+        val tunnisteet: List<KayttajaTunnisteEntity> = kayttajaTunnisteRepository.findAll()
+        val kayttajat: List<HankeKayttajaEntity> = hankeKayttajaRepository.findAll()
+        assertThat(tunnisteet).hasSize(4) // 4 yhteyshenkilo subcontacts.
+        assertThat(kayttajat).hasSize(4)
+        assertThat(tunnisteet).areValid()
+        assertThat(kayttajat).areValid(hanke.id)
+    }
+
+    @Test
+    fun `saveNewTokensFromHanke with pre-existing permissions does not create duplicate`() {
+        val hanke = hankeFactory.save(HankeFactory.create())
```

Comment:
Isn't this the same as just `val hanke = hankeFactory.save()`?

Reply:
Yes, changed it to:

val hanke = hankeFactory.save()